### ISSUE TITLE
Dataplane doesn't reconnect to changing control plane IP

### ISSUE THREAD

**Issue by ludov04** [kumahq/kuma#1620]
### Summary

- It seems that in particular scenarios, some DP don't receive updates anymore from the CP. That result in outages when a service deploys and some of the Kong's DP are not receiving the new IP adresses for that service
- After some digging, we think that it might be related to how the DPs are resolving the CP IP adresses.
- Our CP is using the Service Discovery feature of ECS, like it is documented in the examples, therefore, we have kuma-cp.datacamp-services.datacamp.com dns name that resolve to a (or multiple) A records.
  These records regularly update when the control planes service moves task around, hence the Control Plane IP address regularly changes.
- What we think is happening is that the DP caches the first record it resolves and never resolves it again, and that result in it being disconnected and not receiving updates.
Here is an example from one of these disconnected DP, the current Control Plane IP is `10.3.12.54`
```
/ $ netstat -tpe | grep 5678
tcp        0      0 ip-10-3-12-203.ec2.internal:60034 ip-10-3-12-67.ec2.internal:5678 ESTABLISHED 22/envoy
tcp        0      0 ip-10-3-12-203.ec2.internal:60016 ip-10-3-12-67.ec2.internal:5678 ESTABLISHED 22/envoy
tcp        0      0 ip-10-3-12-203.ec2.internal:60020 ip-10-3-12-67.ec2.internal:5678 ESTABLISHED 22/envoy
tcp        0      0 ip-10-3-12-203.ec2.internal:60008 ip-10-3-12-67.ec2.internal:5678 ESTABLISHED 22/envoy
```
At the same point in time, the output from a healthy dataplane
```
/ $ netstat -tpe | grep 5678
tcp        0      0 ip-10-3-12-66.ec2.internal:35702 ip-10-3-12-54.ec2.internal:5678 ESTABLISHED 22/envoy
tcp        0      0 ip-10-3-12-66.ec2.internal:35696 ip-10-3-12-54.ec2.internal:5678 ESTABLISHED 22/envoy
tcp        0      0 ip-10-3-12-66.ec2.internal:35706 ip-10-3-12-54.ec2.internal:5678 ESTABLISHED 22/envoy
tcp        0      0 ip-10-3-12-66.ec2.internal:35710 ip-10-3-12-54.ec2.internal:5678 ESTABLISHED 22/envoy
```
- We were confused tho of why in the control plantes, these dataplanes kept reporting as "online"
We tried to reproduce it on our staging environment, redeploying the control plane while looking at the netstat output of one of the dataplanes:
The new CP task is running at 10.2.12.150
```
Every 2.0s:  netstat -tpe  | grep 5678                                                                            2021-03-02 15:03:55
tcp        0      0 ip-10-3-11-213.ec2.internal:38594 ip-10-3-12-67.ec2.internal:5678 ESTABLISHED 21/envoy
tcp        0      0 ip-10-3-11-213.ec2.internal:60126 ip-10-3-12-150.ec2.internal:5678 ESTABLISHED 21/envoy
tcp        0      0 ip-10-3-11-213.ec2.internal:38592 ip-10-3-12-67.ec2.internal:5678 ESTABLISHED 21/envoy
tcp        0      0 ip-10-3-11-213.ec2.internal:38584 ip-10-3-12-67.ec2.internal:5678 ESTABLISHED 21/envoy
```
You can see one socket is connected to the correct CP, we killed the CP a couple of times and that one socket keeps connecting back to the right CP.
But looking at the `/config_dump`, it was not in sync with the newest configuration, and you can see some of the socket still report being connected to the old CP IP.
The control plane is currently behind a network load balancer for exposing the DNS port, but the dataplane are connecting directly to it by resolving the A record.

### Steps To Reproduce
In Universal,
1. Have the Control Plane DNS name resolve to A records (we are using the service discovery feature from ECS, which update these records whenever the tasks change)
2. Spin up new dataplanes
3. Redeploy control plane => dns name should resolve to different IPs
4. Dataplane doesn't receive updates from control plane.

### Additional Details & Logs

- Version: 1.0.8
- Error logs
- Configuration
Envoy:
```
{
  "params": {
    "Id": "datacamp-services.careerhub-api-39e402ef0a22434e98c21adaac133c88",
    "Service": "careerhub-api",
    "AdminAddress": "127.0.0.1",
    "AdminPort": 9901,
    "AdminAccessLogPath": "/dev/null",
    "XdsHost": "kuma-cp.datacamp-services.datacamp.com",
    "XdsPort": 5678,
    "XdsConnectTimeout": 1000000000,
    "AccessLogPipe": "/tmp/kuma-al-careerhub-api-39e402ef0a22434e98c21adaac133c88-datacamp-services.sock",
    "DataplaneTokenPath": "/tmp/kuma-dp-817797992/careerhub-api-39e402ef0a22434e98c21adaac133c88",
    "DataplaneResource": "{\"type\":\"Dataplane\",\"mesh\":\"datacamp-services\",\"name\":\"careerhub-api-39e402ef0a22434e98c21adaac133c88\",\"creationTime\":\"0001-01-01T00:00:00Z\",\"modificationTime\":\"0001-01-01T00:00:00Z\",\"networking\":{\"address\":\"10.4.11.29\",\"inbound\":[{\"port\":15000,\"servicePort\":3000,\"tags\":{\"ecs/task_id\":\"39e402ef0a22434e98c21adaac133c88\",\"ecs/task_revision\":\"42\",\"kuma.io/protocol\":\"http\",\"kuma.io/service\":\"careerhub-api\"}}],\"transparentProxying\":{\"redirectPortInbound\":15006,\"redirectPortOutbound\":15001}}}",
    "CertBytes": "REDACTED",
    "KumaDpVersion": "1.0.8",
    "KumaDpGitTag": "1.0.8",
    "KumaDpGitCommit": "e099dcdf6ab0af4ae5c2c003e234bf0f72901a75",
    "KumaDpBuildDate": "2021-02-19T13:24:22Z",
    "EnvoyVersion": "1.16.2",
    "EnvoyBuild": "e98e41a8e168af7acae8079fc0cd68155f699aa3/1.16.2/Clean/RELEASE/BoringSSL",
    "HdsEnabled": true
  }
}
```
Kuma:
```
{
  "apiServer": {
    "auth": {
      "allowFromLocalhost": true,
      "clientCertsDir": "/kuma/certs/clients"
    },
    "catalog": {
      "bootstrap": {
        "url": ""
      }
    },
    "corsAllowedDomains": [
      ".*"
    ],
    "http": {
      "enabled": true,
      "interface": "0.0.0.0",
      "port": 5681
    },
    "https": {
      "enabled": true,
      "interface": "0.0.0.0",
      "port": 5682,
      "tlsCertFile": "/kuma/certs/server.pem",
      "tlsKeyFile": "/kuma/keys/server.pem"
    },
    "readOnly": false
  },
  "bootstrapServer": {
    "apiVersion": "v2",
    "params": {
      "adminAccessLogPath": "/dev/null",
      "adminAddress": "127.0.0.1",
      "adminPort": 0,
      "xdsConnectTimeout": "1s",
      "xdsHost": "",
      "xdsPort": 5678
    }
  },
  "defaults": {
    "skipMeshCreation": false
  },
  "diagnostics": {
    "debugEndpoints": false,
    "serverPort": 5680
  },
  "dnsServer": {
    "CIDR": "240.0.0.0/4",
    "domain": "mesh",
    "port": 5653
  },
  "dpServer": {
    "auth": {
      "type": "dpToken"
    },
    "hds": {
      "checkDefaults": {
        "healthyThreshold": 1,
        "interval": "1s",
        "noTrafficInterval": "1s",
        "timeout": "2s",
        "unhealthyThreshold": 1
      },
      "enabled": true,
      "interval": "5s",
      "refreshInterval": "10s"
    },
    "port": 5678,
    "tlsCertFile": "/kuma/certs/server.pem",
    "tlsKeyFile": "/kuma/keys/server.pem"
  },
  "environment": "universal",
  "general": {
    "dnsCacheTTL": "10s",
    "tlsCertFile": "/tmp/305653175.crt",
    "tlsKeyFile": "/tmp/093040042.key"
  },
  "guiServer": {
    "apiServerUrl": "http://kuma-cp.datacamp-services.datacamp.com:5681/"
  },
  "metrics": {
    "dataplane": {
      "enabled": true,
      "subscriptionLimit": 10
    },
    "mesh": {
      "maxResyncTimeout": "20s",
      "minResyncTimeout": "1s"
    },
    "zone": {
      "enabled": true,
      "subscriptionLimit": 10
    }
  },
  "mode": "standalone",
  "monitoringAssignmentServer": {
    "assignmentRefreshInterval": "1s",
    "grpcPort": 5676
  },
  "multizone": {
    "global": {
      "kds": {
        "grpcPort": 5685,
        "refreshInterval": "1s",
        "tlsCertFile": "/tmp/305653175.crt",
        "tlsKeyFile": "/tmp/093040042.key",
        "zoneInsightFlushInterval": "10s"
      },
      "pollTimeout": "500ms"
    },
    "remote": {
      "kds": {
        "refreshInterval": "1s",
        "rootCaFile": ""
      }
    }
  },
  "reports": {
    "enabled": true
  },
  "runtime": {
    "kubernetes": {
      "admissionServer": {
        "address": "",
        "certDir": "",
        "port": 5443
      },
      "controlPlaneServiceName": "kuma-control-plane",
      "injector": {
        "caCertFile": "",
        "cniEnabled": false,
        "exceptions": {
          "labels": {
            "openshift.io/build.name": "*",
            "openshift.io/deployer-pod-for.name": "*"
          }
        },
        "initContainer": {
          "image": "kuma/kuma-init:latest"
        },
        "sidecarContainer": {
          "adminPort": 9901,
          "drainTime": "30s",
          "gid": 5678,
          "image": "kuma/kuma-dp:latest",
          "livenessProbe": {
            "failureThreshold": 12,
            "initialDelaySeconds": 60,
            "periodSeconds": 5,
            "timeoutSeconds": 3
          },
          "readinessProbe": {
            "failureThreshold": 12,
            "initialDelaySeconds": 1,
            "periodSeconds": 5,
            "successThreshold": 1,
            "timeoutSeconds": 3
          },
          "redirectPortInbound": 15006,
          "redirectPortOutbound": 15001,
          "resources": {
            "limits": {
              "cpu": "1000m",
              "memory": "512Mi"
            },
            "requests": {
              "cpu": "50m",
              "memory": "64Mi"
            }
          },
          "uid": 5678
        },
        "sidecarTraffic": {
          "excludeInboundPorts": [],
          "excludeOutboundPorts": []
        },
        "virtualProbesEnabled": true,
        "virtualProbesPort": 9000
      },
      "marshalingCacheExpirationTime": "5m0s"
    },
    "universal": {
      "dataplaneCleanupAge": "72h0m0s"
    }
  },
  "sdsServer": {
    "dataplaneConfigurationRefreshInterval": "1s"
  },
  "store": {
    "cache": {
      "enabled": true,
      "expirationTime": "1s"
    },
    "kubernetes": {
      "systemNamespace": "kuma-system"
    },
    "postgres": {
      "connectionTimeout": 5,
      "dbName": "kuma",
      "host": "datacamp-services-psql-shared.crsd2orwscer.us-east-1.rds.amazonaws.com",
      "maxOpenConnections": 0,
      "maxReconnectInterval": "1m0s",
      "minReconnectInterval": "10s",
      "password": "*****",
      "port": 5432,
      "tls": {
        "caPath": "",
        "certPath": "",
        "keyPath": "",
        "mode": "disable"
      },
      "user": "kuma"
    },
    "type": "postgres",
    "upsert": {
      "conflictRetryBaseBackoff": "100ms",
      "conflictRetryMaxTimes": 5
    }
  },
  "xdsServer": {
    "dataplaneConfigurationRefreshInterval": "1s",
    "dataplaneStatusFlushInterval": "10s"
  }
}
```
- Platform and Operating System: Linux / ECS

**Comment by jakubdyszkiewicz** [comment]
Hey,

thank you for such a detailed report.
I was able to reproduce a problem. I think we did not catch it because we were always deploying it as 1 instance or behind an LB which had 1 IP or on Kubernetes which has also 1 IP (ClusterIP of Service).

It seems like a problem in Envoy itself https://github.com/envoyproxy/envoy/issues/15380 I'm in touch with the community to figure out if this is expected or not.

I have one workaround to implement in mind, but I'm holding off for now since it's a bit less secure option.

I'll keep you posted about this.
